src/vellum/workflows/runner/runner.py

```diff
@@ -250,6 +258,105 @@ def __init__(
+        """Get all entrypoints connected to a specific trigger type.
+
+        Allows subclasses: if trigger_class is a subclass of any declared trigger,
+        returns those entrypoints.
+        """
+        entrypoints: List[Type[BaseNode]] = []
+        for subgraph in self.workflow.get_subgraphs():
+            for trigger in subgraph.triggers:
+                # Check if the provided trigger_class is a subclass of the declared trigger
+                # This allows runtime instances to be subclasses of what's declared in the workflow
+                if issubclass(trigger_class, trigger):
+                    entrypoints.extend(subgraph.entrypoints)
+        return entrypoints
+
+    def _process_trigger(self, trigger: Optional[IntegrationTrigger]) -> None:
+        """
+        Process trigger and bind trigger attributes to state.
+
+        This method handles validation and routing for workflows with IntegrationTrigger:
+        - When trigger is provided: validates compatibility, binds to state, and routes to the
+          specific trigger path (respects caller-provided entrypoints)
+        - When trigger is missing but workflow has IntegrationTrigger: validates workflow can run
+          without it (requires ManualTrigger) and routes to ManualTrigger path
+        - When no trigger and no IntegrationTrigger: default ManualTrigger behavior (backward compatibility)
+
+        Entrypoint filtering behavior:
+        - ONLY filters entrypoints when caller hasn't explicitly provided them via entrypoint_nodes or external_inputs
+        - Preserves backward compatibility for existing code
+        """
+        # Collect all IntegrationTrigger types declared in the workflow
+        workflow_integration_triggers = []
+        for subgraph in self.workflow.get_subgraphs():
+            for trigger_type in subgraph.triggers:
+                if issubclass(trigger_type, IntegrationTrigger):
+                    workflow_integration_triggers.append(trigger_type)
+
+        # Case 1: Trigger instance provided - validate and bind
+        if trigger:
+            if not workflow_integration_triggers:
+                raise WorkflowInitializationException(
+                    message="trigger provided but workflow does not have an IntegrationTrigger",
+                    workflow_definition=self.workflow.__class__,
+                    code=WorkflowErrorCode.INVALID_INPUTS,
+                )
+
+            # Validate that the trigger instance is compatible with one of the workflow's trigger types
+            # Allow subclasses: if workflow declares BaseSlackTrigger, accept SpecificSlackTrigger instances
+            trigger_class = type(trigger)
+
+            # Check if the provided trigger_class is a subclass of any declared trigger type
+            if not any(
+                issubclass(trigger_class, declared_trigger) for declared_trigger in workflow_integration_triggers
+            ):
```

Comment:
[nit] feels like we could use the initial loop in L301-L304 to just search or the trigger we want to invoke, instead of maintain an array of them through here